src/hdmf/common/resources.py

```diff
@@ -0,0 +1,355 @@
+        {'name': 'resources', 'child': True},
+        {'name': 'objects', 'child': True},
+        {'name': 'object_keys', 'child': True},
+    )
+
+    @docval({'name': 'name', 'type': str, 'doc': 'the name of this ExternalResources container'},
+            {'name': 'keys', 'type': KeyTable, 'default': None,
+             'doc': 'the table storing user keys for referencing resources'},
+            {'name': 'resources', 'type': ResourceTable, 'default': None,
+             'doc': 'the table storing resource information'},
+            {'name': 'objects', 'type': ObjectTable, 'default': None,
+             'doc': 'the table storing object information'},
+            {'name': 'object_keys', 'type': ObjectKeyTable, 'default': None,
+             'doc': 'the table storing object-resource relationships'})
+    def __init__(self, **kwargs):
+        name = popargs('name', kwargs)
+        super().__init__(name)
+        self.keys = kwargs['keys'] or KeyTable()
+        self.resources = kwargs['resources'] or ResourceTable()
+        self.objects = kwargs['objects'] or ObjectTable()
+        self.object_keys = kwargs['object_keys'] or ObjectKeyTable()
+
+    def add_key(self, key):
+        """
+        Add a key to be used for making references to external resources
+
+        It is possible to use the same *key_name* to refer to different resources so long as the *key_name* is not
+        used within the same object and field. To do so, this method must be called for the two different resources.
+        The returned Key objects must be managed by the caller so as to be appropriately passed to subsequent calls
+        to methods for storing information about the different resources.
+        """
+        return Key(key, table=self.keys)
+
+    @docval({'name': 'key', 'type': (str, Key), 'doc': 'the key to associate the resource with'},
+            {'name': 'resource_name', 'type': str, 'doc': 'the name of the resource'},
+            {'name': 'resource_entity_id', 'type': str, 'doc': 'the entity at the resource to associate'},
+            {'name': 'resource_entity_uri', 'type': str, 'doc': 'the URL for the entity at the resource'})
+    def add_resource(self, **kwargs):
+        """
+        Add an external resource that will be referenced to using the given key
+        """
+        key = kwargs['key']
+        resource_name = kwargs['resource_name']
+        resource_entity_id = kwargs['resource_entity_id']
+        resource_entity_uri = kwargs['resource_entity_uri']
+        if not isinstance(key, Key):
+            key = self.add_key(key)
+        resource_entity = ResourceEntity(key, resource_name, resource_entity_id, resource_entity_uri,
+                                         table=self.resources)
+        return resource_entity
+
+    @docval({'name': 'container', 'type': (Container, str), 'doc': 'the Container to add'},
+            {'name': 'field', 'type': str, 'doc': 'the field on the Container to add'})
```

Comment:
```suggestion
            {'name': 'field', 'type': str, 'doc': 'the field on the container to add', 'default': None})
```

Reply:
The `field` will be tricky because fields can be remapped between the schema and the API.... The `field` here should specify the schema field, which the user does not yet have access to, I believe.

Relatedly, `Data` objects have a data field that does not map to a field in the schema, in which case the `field` should be None or empty string.

Reply:
Thinking this through more, I think the `field` here should be the API class attribute name, and this function should map the `field` name to the schema name and store that.